pyfive/dataobjects.py

```diff
@@ -722,6 +733,11 @@ def determine_data_shape(buf, offset):
         offset += DATASPACE_MSG_HEADER_V2_SIZE
     else:
         raise InvalidHDF5File('unknown dataspace message version')
+    # to detect Empty aka NULL dataspace we need to check for V2 DATASPACE MESSAGE
+    if header["version"] == 2:
+        # check for NULL dataspace and return early
+        if header["type"] == 2:
+            return None, None
 
```

Comment:
That's the crucial part, see https://support.hdfgroup.org/documentation/hdf5/latest/_f_m_t2.html#subsubsec_fmt2_dataobject_hdr_msg_simple section V2 for details.

Reply:
I'd be tempted to move the `if header['type]] == 2; return None, None` into the version elif ... but I'm also happy to leave it here.